### DISCUSSION TITLE
How to create a create custom decorator?

### DISCUSSION THREAD

**Discussion by kiragengis** [allure-framework/allure2#2382] [Questions & Support] [answered]
With the help of @delatrie, I have created a custom function to display the returning value of a function in the report. The solution is to create a helper function like this one:
```
public static T Step<T>(string name, Func<T> func) => AllureApi.Step(name, () =>
{
    var returnValue = func();
    AllureLifecycle.Instance.UpdateStep(s => s.parameters.Add(new()
    {
        name = "returnValue",
        value = returnValue.ToString()
    }));
    return returnValue;
});
```
Then use it like this:
```
[Test]
public void Test1()
{
    Step("My custom step", () => MyMethod("Testing from Test1"));
}

private string MyMethod(string test)
{    
    return test;
}
```

What I want to do now is to create a custom decorator similar to AllureStep to avoid calling my custom function and decorate my methods with it. Is that possible?

It would look something like:
```
[MyCustomAllureStep("A private method with the parameter: {test}")]
private string MyMethod(string test)
{    
    return test;
}
```

**Comment by delatrie** [comment, marked as answer]
Absolutely. We use [Aspect Injector](https://github.com/pamidur/aspect-injector) to wrap methods in steps. You're not limited to that particular AOP tool, but it would be the most straightforward approach to define a custom step attribute with an extra injection.

First, define an aspect that does the work of adding an extra parameter. I advise you to use `Kind.After` as it supports async methods without extra effort (both `Task<T>` and `ValueTask<T>` return types are supported out of the box):

```c#
[Aspect(Scope.Global)]
public class AddReturnValueAspect
{
    [Advice(Kind.After)]
    public void AddReturnValueToAllureStep(
        [Argument(Source.ReturnValue)] object? returnValue
    )
    {
        AllureLifecycle.Instance.UpdateTestCase(
            t => t.steps.Last().parameters.Add(new()
            {
                name = "returnValue",
                value = returnValue?.ToString() ?? "null"
            })
        );
    }
}
```

You might wonder why use `UpdateTestCase` instead of `UpdateStep`. The reason is that the effect that actually adds an Allure step is of a `Kind.Around` type and hence is always weaved before effects of other kinds making the step inaccessible to them. But for the exact same reason, it's guaranteed that in a `Kind.After` effect the last step of the current test is the exact step we need to add the parameter to.

Then, define a custom attribute that you will apply instead of `[AllureStep]`:

```c#
[Injection(typeof(AddReturnValueAspect))]
public class StepWithValueAttribute : AllureStepAttribute
{
    public StepWithValueAttribute(string? name = null) : base(name) { }
}
```

Now, apply the attribute to methods:

```c#
public class CustomStepTests
{
    [Test]
    public void SyncTestWithStep()
    {
        SyncStep();
    }

    [Test]
    public async Task AsyncTestWithStep()
    {
        await AsyncStep();
    }

    [StepWithValue("A sync step")]
    static string SyncStep()
    {
        return "Hello from a synchronous step";
    }

    [StepWithValue("An async step")]
    static async ValueTask<int> AsyncStep()
    {
        return await ValueTask.FromResult(25);
    }
}
```

**Reply by kiragengis** [reply]
Thank you @delatrie ! Everything works fine for steps that are not nested. When I use nested steps, this is the result:

![image](https://github.com/allure-framework/allure2/assets/33094881/d0243763-aad0-4d18-a864-00b86229caad)


**This is my code:**
```
[Test]
public void Test100()
{
    MethodWithCustomParameter(new WhoAmI("string from test", 15), "This is the second");
}

[StepWithValue("A sync step")]
private string MethodWithCustomParameter([Name("Custom class object")] WhoAmI test, [Name("Second parameter that is a string")] string secondParameter)
{
    InnerStep(test, secondParameter);
    return "WhoAmIReturning";
}

[StepWithValue("A sync step")]
private string InnerStep([Name("Inner Custom class object")] WhoAmI test, [Name("Inner Second parameter that is a string")] string secondParameter)
{
    return "This is my inner step result";
}
```

Am I missing something? Thank you again for your help

**Reply by kiragengis** [reply]
I can see the problem is in this line:
```
AllureLifecycle.Instance.UpdateTestCase(
    t => t.steps.Last().parameters.Add(new()
    {
        name = "returnValue",
        value = returnValue?.ToString() ?? "null"
    })
);
```

I am not reading the sub-steps. I must find a way to navigate through all the children/descendants and get the latest.

**Reply by delatrie** [reply]
Oh, it's a nice catch. What you actually want to do is to navigate down to the last step until you find the first one that is finished:
```c#
static StepResult GetLastStep(TestResult t)
{
    var step = t.steps.Last();
    while (step.stage != Stage.finished)
    {
        step = step.steps.Last();
    }
    return step;
}
```

Replace `t.steps.Last()` in the initial example with `GetLastStep(t)` and that should do the trick.

**Reply by kiragengis** [reply]
That fixed my problem. Thank you @delatrie !
